balancing/balance_breaker.go

```diff
@@ -0,0 +1,541 @@
+	Record(duration time.Duration, success bool) bool
+	ShouldOpen() bool
+}
+
+func newBreaker(retention int, callTimeLimit time.Duration,
+	timeLimitPercentile, errorRate float64,
+	closeDelay, maxDelay time.Duration) Breaker {
+
+	return &NodeBreaker{
+		timeData:            newLenLimitCounter(retention),
+		successData:         newLenLimitCounter(retention),
+		rate:                errorRate,
+		callTimeLimit:       callTimeLimit,
+		timeLimitPercentile: timeLimitPercentile,
+		now:                 time.Now,
+		closeDelay:          closeDelay,
+		maxDelay:            maxDelay,
+	}
+}
+
+// NodeBreaker is implementation of Breaker interface
+type NodeBreaker struct {
+	rate                float64
+	callTimeLimit       time.Duration
+	timeLimitPercentile float64
+	timeData            *lenLimitCounter
+	successData         *lenLimitCounter
+	now                 func() time.Time
+	closeDelay          time.Duration
+	maxDelay            time.Duration
+	isOpen              bool
+	state               *openStateTracker
+}
+
+// Record collects call data and returns bool if breaker should be open
+func (breaker *NodeBreaker) Record(duration time.Duration, success bool) bool {
+	breaker.timeData.Add(float64(duration))
+	successValue := float64(1)
+	if success {
+		successValue = float64(0)
+	}
+	breaker.successData.Add(successValue)
+	return breaker.ShouldOpen()
+}
+
+// ShouldOpen checks if breaker should be open
+func (breaker *NodeBreaker) ShouldOpen() bool {
+	exceeded := breaker.limitsExceeded()
+	if breaker.state != nil {
+		return breaker.checkStateHalfOpen(exceeded)
+	}
+
+	if exceeded {
+		breaker.openBreaker()
+	}
+	return exceeded
+}
+
+func (breaker *NodeBreaker) checkStateHalfOpen(exceeded bool) bool {
+	state, changed := breaker.state.currentState(breaker.now(), exceeded)
+	if state == closed {
+		if changed {
+			breaker.state = nil
+		}
+		return false
+	}
+	if state == halfopen {
+		if changed {
+			breaker.reset()
+		}
+		return false
+	}
+	return true
+}
+
+func (breaker *NodeBreaker) limitsExceeded() bool {
+	if breaker.errorRate() > breaker.rate {
+		breaker.openBreaker()
+		return true
+	}
+
+	if breaker.timeData.Percentile(breaker.timeLimitPercentile) > float64(breaker.callTimeLimit) {
+		breaker.openBreaker()
+		return true
+	}
+	return false
+}
+
+func (breaker *NodeBreaker) openBreaker() {
+	if breaker.state != nil {
+		return
+	}
+	breaker.state = newOpenStateTracker(
+		breaker.now(), breaker.closeDelay, breaker.maxDelay)
+}
+
+func (breaker *NodeBreaker) reset() {
+	breaker.timeData.Reset()
+	breaker.successData.Reset()
+}
+
+func (breaker *NodeBreaker) errorRate() float64 {
+	sum := breaker.successData.Sum()
+	count := float64(len(breaker.successData.values))
+	return sum / count
+}
+
+func newLenLimitCounter(retention int) *lenLimitCounter {
+	return &lenLimitCounter{
+		values: make([]float64, retention, retention),
+	}
+}
+
+type lenLimitCounter struct {
+	values  []float64
+	nextIdx int
+	mx      sync.Mutex
```

Comment:
the structure has a mutex as a member, but none of it's metods actually use it - `Add` etc.